### DISCUSSION TITLE
Setting Focus on a Formik Field Dynamically

### DISCUSSION THREAD

**Discussion by oller** [jaredpalmer/formik#3396] [Q&A]
Hello all,

I'm interested to hear how one might best approach this requirement.  I can't find much in the docs related to this.  

Let's say I want to deep link into a form with a specific field focussed, this needs to be dynamic, so if you click on "Button A", then "Input A" is focussed, another button and input b would be.  I can pass some meta data (i thought the field.name I want focussed) through the route, and have access to that when I initialise the form, but it'd be great as opposed to conditionally handling this with each `<Field>` element, if there was a way to wrap formik or have a prop at the `<Formik>` level.

i.e.

```
<Formik focussedField={someConstMatchingField.Name} ... >
```

Any input & insight appreciated!

**Comment by johnrom** [comment]
Interesting question. Formik doesn't provide functionality for this, but this is how I would go about it. You can simply ignore the TS if you don't use it.

Focusability would be a useful [Plugin](https://github.com/formium/formik/issues/3109)

```tsx
import { FieldAttributes } from 'formik';

interface FocusableFieldProps extends FieldAttributes  {
  onFocusRef: React.MutableRefObject<Fn>;
}

const FocusableField = ({ onFocusRef, innerRef, ...fieldProps }: FocusableFieldProps) => {
  // Field accepts innerRef, we'll check if it exists and if it doesn't we'll use our own.
  // This only happens once on mount so don't switch refs between renders.

  // internally it's very ugly, but that's why we are abstracting it to make it relatively easy to use.
  const backupRef = useRef<HTMLElement>(null);
  // it's a ref of a ref, use .current.current
  const ref = useRef(innerRef ?? backupRef);

  const handleFocus = useCallback(() => {
    ref.current.current?.focus();
  }, []);

  useEffect(() => {
    props.onFocusRef?.current = handleFocus;
  }, []);

  // Field accepts innerRef
  return <Field innerRef={ref} {...fieldProps} />
}

const MyForm = () => {
  const firstFocusRef = useRef(null);
  const secondFocusRef = useRef(null);

  return <Formik {...props}>
    <a onClick={() => firstFocusRef.current?.()}>Focus First Field</a>
    <a onClick={() => secondFocusRef.current?.()}>Focus Second Field</a>
    <FocusableField name="my-first-field" onFocusRef={myFirstFocusRef} />
    <FocusableField name="my-second-field" onFocusRef={mySecondFocusRef} />
  </Formik>
}
```

**Reply by oller** [reply]
Heya, thanks for the response! Would this work if the `onClick` events were fired in another component, so all the above component has to go on, is say a prop coming through with a string of the field we want to apply focus to?

Ideally any solution would be abstracted to the point It could be easily applied to any one of our fomik forms.  Thanks again!

**Reply by johnrom** [reply]
No, but you don't need Formik for that, you just need the ID of your input.

```tsx
interface FocusInputLinkProps {
  id: string;
}
const FocusFieldLink = ({ id, children }: PropsWithChildren<FocusInputLinkProps>) => {
  return <button 
    onClick={(event) => {
      event.preventDefault();
  
      document.querySelector(`#${id}`})?.focus();
    }}
  >{children}</button>
};

const MyForm = () => {
  return <Formik {...props}>
    <Form>
      <Field name="firstName" id="firstName" />
    </Form>
  </Formik>
}

const App = () => {
  return <>
    <MyForm />
    <FocusFieldLink id="firstName">Focus your Field</FocusFieldLink>
  </>
}
```
